File: platform/extension-sdk/node/src/platform-client.ts
import { createDefaultCredential, type TokenCredentialLike } from "./credential.js";
import { createKeyVaultClient, type KeyVaultClient } from "./key-vault.js";
import { createPlatformConfigFromEnv, type PlatformConfig } from "./platform-config.js";
import { createSqlDatabaseClient, type SqlDatabaseClient } from "./sql-database.js";
import { configurePlatformTelemetry, type PlatformTelemetryClient } from "./telemetry.js";

export interface PlatformClientOptions {
  config?: Partial<PlatformConfig>;
  credential?: TokenCredentialLike;
  telemetryEnabled?: boolean;
}

export interface PlatformClient {
  config: PlatformConfig;
  credential: TokenCredentialLike;
  keyVault?: KeyVaultClient;
  sqlDatabase?: SqlDatabaseClient;
  initializeTelemetry(): Promise<PlatformTelemetryClient>;
}

export async function createPlatformClient(options: PlatformClientOptions = {}): Promise<PlatformClient> {
  const config = createPlatformConfigFromEnv(options.config);
  const credential = options.credential ?? await createDefaultCredential();
  const keyVault = config.keyVaultUrl
    ? await createKeyVaultClient({ vaultUrl: config.keyVaultUrl, credential })
    : undefined;
  const sqlDatabase = config.sqlServerHost && config.sqlDatabaseName
    ? await createSqlDatabaseClient({
      server: config.sqlServerHost,
      database: config.sqlDatabaseName,
      credential
    })
    : undefined;

  return {
    config,
    credential,
    keyVault,
    sqlDatabase,
    async initializeTelemetry(): Promise<PlatformTelemetryClient> {
      return configurePlatformTelemetry({
        serviceName: config.serviceName,
        connectionString: config.applicationInsightsConnectionString,
        enabled: options.telemetryEnabled
      });
    }
  };
}

File: platform/extension-sdk/python/src/aks_platform_sdk/platform_client.py
from dataclasses import dataclass
from typing import Any

from .credential import create_default_credential
from .key_vault import KeyVaultClient
from .platform_config import PlatformConfig, create_platform_config_from_env
from .sql_database import SqlDatabaseClient
from .telemetry import PlatformTelemetryClient, configure_platform_telemetry


@dataclass
class PlatformClient:
    config: PlatformConfig
    credential: Any
    key_vault: KeyVaultClient | None = None
    sql_database: SqlDatabaseClient | None = None

    def initialize_telemetry(self, enabled: bool = True) -> PlatformTelemetryClient:
        return configure_platform_telemetry(
            service_name=self.config.service_name,
            connection_string=self.config.application_insights_connection_string,
            enabled=enabled,
        )


def create_platform_client(config: PlatformConfig | None = None, credential: Any | None = None) -> PlatformClient:
    resolved_config = config or create_platform_config_from_env()
    resolved_credential = credential or create_default_credential()

    key_vault = (
        KeyVaultClient(resolved_config.key_vault_url, resolved_credential)
        if resolved_config.key_vault_url
        else None
    )
    sql_database = (
        SqlDatabaseClient(resolved_config.sql_server_host, resolved_config.sql_database_name, resolved_credential)
        if resolved_config.sql_server_host and resolved_config.sql_database_name
        else None
    )

    return PlatformClient(
        config=resolved_config,
        credential=resolved_credential,
        key_vault=key_vault,
        sql_database=sql_database,
    )
